Implement Function: Array.pick_random()

nonunknown opened the pull request:
Implements: https://github.com/godotengine/godot-proposals/issues/3454

This is partially implemented, since some users suggested adding for another classes, but I think each class should have their own PR, what you guys think?

**IMPORTANT:**
First merge #54940 before this one, to avoid https://github.com/godotengine/godot/issues/54939

test code:
```gdscript
var arr:Array = []

func _ready() -> void:
	arr.append_array(["test",1,2,3])
	print(arr)
	for i in range(10):
		print("Result: ", arr.pick_random())
```

output:

**First Run**:
```
[test, 1, 2, 3]
Result: 2
Result: 1
Result: 3
Result: 1
Result: 3
Result: 3
Result: 2
Result: 1
Result: 1
Result: 1

```

**Second Run**:
```
[test, 1, 2, 3]
Result: 1
Result: test
Result: 2
Result: 2
Result: 1
Result: 1
Result: 2
Result: 3
Result: 3
Result: 2

```

naithar committed:
[iOS] Extend iOS plugins to support Swift runtime

Calinou committed:
Use interleaved gradient noise for distance fade

This provides better visuals and performance compared to the previous
ordered dither.

raulsntos committed:
Fix typo in weakref documentation

Calinou committed:
Tweak PopupMenu item spacing in the editor theme

- Increase spacing between items for easier clicking with the mouse.
- Increase lateral margins for better visual appearance.
- Decrease margin at the top and bottom to compensate for the increased
  per-item height.

sairam4123 commented:
Would be interested in adding in probabilities too?
Like this?
`[1, 0, 2, 3].pick_random([0.25, 0.25, 0.25, 0.25])`
It's your choice to to decide whether the sum of all probabilities must be 1. It's how it works in Python.

nonunknown commented:
> Would be interested in adding in probabilities too? Like this? `[1, 0, 2, 3].pick_random([0.25, 0.25, 0.25, 0.25])` It's your choice to to decide whether the sum of all probabilities must be 1. It's how it works in Python.

please create a proposal for that, every implementation depends on the users needs.

rcorre committed:
Migrate project list to separate config file.

Storing the project/favorites list in the EditorSettings makes it
difficult to version-control your editor configuration, as the file will
continually change as you open new projects. It also means a
configuration can't be shared across machines, as they might not have
the same projects or file layout.

Now the project list is stored in $godot_data_dir/projects.cfg.
Each path is a section, which has a boolean favorite value.
If the new config does not exist, the editor attempts to migrate legacy
EditorSettings-based configuration to the new file.

Fixes godotengine/godot-proposals#1637.

raulsntos committed:
Add `@GDScript.type_exists` documentation

Calinou committed:
Rename shader parameter uniform setter/getter methods for consistency

`shader_uniform` is now consistenly used across both per-shader
and per-instance shader uniform methods. This makes methods easier
to find in the class reference when looking for them.

Calinou committed:
Improve warning message for invalid desktop VRAM compression settings

- Display the source file path.
- Display the valid compression algorithms on desktop platforms.
- Use "desktop platforms" instead of "PC" to avoid ambiguity.
- Use "may" instead of "will", as desktop platforms can still sometimes
  display ETC1/ETC2 textures depending on the GPU and drivers.

Geometror committed:
Add "Make Unique (Recursive)" to resource picker, simplify "Make Unique"

SaracenOne committed:
Disable plugin for conversion of foreign Sprite2D to different types

SaracenOne committed:
Add read-only mode to AnimationEditor for foreign resources.

hoontee committed:
Add `TorusMesh`

KoBeWi committed:
Restore old lerp() behavior and add lerpf()

iwek7 committed:
Fix issue with overlaping text label text

Calinou committed:
Add a shadow opacity property to Light3D

This can be used to make shadows translucent for a specific light.

The light distance fade system also uses this to smoothly fade the shadow
when the light fade transition distance is greater than 0.

Step-an committed:
Add drag and drop support to EditorPropertyPath

clayjohn committed:
Evaluate specular reflections using specular dominant direction instead of assuming mirror reflections

akien-mga reviewed (approved):
Approved in proposals review meeting.

akien-mga commented on core/variant/array.h:
```diff
@@ -80,6 +80,7 @@ class Array {
 
 	Variant front() const;
 	Variant back() const;
+	Variant pick_random() const;
```
Trailing whitespace.

akien-mga commented:
You need to update the class reference with `--doctool` and write documentation for the new method: https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html

Apparently there's also a style issue but not sure which, you should likely run clang-format.

Calinou committed:
Add a Bake Mask property to GPUParticlesCollisionSDF3D

This allows not accounting for certain visible meshes during baking
(such as foliage and thin fixtures).

This also adds a clarification about transparent materials always
being excluded in the OccluderInstance3D documentation.

aaronfranke committed:
Remove duplicate data structure for export features

RandomShaper committed:
Set default resource names under pure debug in Vulkan RD

RandomShaper committed:
spirv-reflect: Update to latest commit

samdze committed:
Make Image.get_size() return a Vector2i instead of a Vector2

fire-forge committed:
Use a SpinBox in the inspector array resize dialog

madmiraal committed:
Fix incorrect Android scancodes

aaronfranke committed:
Allow exporting PCK files without export templates installed

fire-forge committed:
Use menu icon for layers grid button

Calinou committed:
Print expected `os.arch` tuple for current platform in GDExtension error

This also adds `Engine.get_architecture_name()` to get the name of the
CPU architecture the Godot binary was built for.

KoBeWi committed:
Split ceil(), floor() and round() method

timothyqiu committed:
Add a Window visibility toggle button to the Scene tree

fire-forge committed:
Fix code font usage in EditorPropertyMultilineText
- Only use code font while in expression mode
- Use code font and syntax highlighter in big text box

dsnopek committed:
Fix locale resource remapping with binary conversion on export

PrecisionRender committed:
Add `operator*` to `Plane`

YeldhamDev committed:
Hide method filter bar when selecting a doc in the Script editor

object71 committed:
Removed faulty function update after get_property_list.

The function tried to rearrange properties but that lead to problems with duplication or deleted properties. Implemented the logic that that function did inside the get_property_list both for tool scripts and non-tool scripts.

qarmin committed:
Various converter enhancements and bugfixes

fire-forge committed:
Improve ShapeCast2D editor and debug drawing
- Rename RayCast2DEditorPlugin to Cast2DEditorPlugin and make it also support editing ShapeCast2D.
- Apply RayCast2D debug drawing improvements from #46675 to ShapeCast2D.

bruvzg committed:
[TextServer] Add a build warning when building with external FreeType without Brotli support.

Rindbee committed:
Fix Hue slider incorrectly rotated 90 degrees in ColorPicker

fire-forge committed:
Adjust 2 default theme colors in ItemList and Tree
- Change Tree's drop_position_color from orange to white.
- Change ItemList's guide_color to the same color as Tree's, which fits in with a dark theme better.

bruvzg committed:
Add Text Server related options to the build profiles editor.

Adds SCons options to disable Brotli and Graphite.
Adds option categories to the build profiles editor.
Adds options default state to the build profiles editor.
Adds Text Server related options to the build profiles editor.
Fix misplaced OpenGL/Vulkan SCons options.

Calinou committed:
Document editor settings in the class reference

RedMser committed:
Implement `keep_screen_on` for Windows

nonunknown commented:
@akien-mga just fixed everything!

akien-mga commented on doc/classes/Array.xml:
```diff
@@ -395,6 +395,17 @@
+			<return type="Variant" />
+			<description>
+				Returns a random value from the target array.
+				[codeblock]
+				var array:Array[int] = [1,2,3,4]
+				print(array.pick_random())
+				# output: 3
+				[/codeblock]
```
```suggestion
				[codeblock]
				var array: Array[int] = [1, 2, 3, 4]
				print(array.pick_random())  # Prints either of the four numbers.
				[/codeblock]
```

akien-mga commented on core/variant/array.cpp:
```diff
@@ -31,6 +31,7 @@
 #include "array.h"
 
 #include "container_type_validate.h"
+#include "core/math/random_number_generator.h"
```
You're not actually using this header here. `Math::rand()` is from `core/math/math_funcs.h`.

akien-mga commented:
Please squash the commits when it's ready, we don't want 5 extra merge commits in the Git history.

akien-mga committed:
Tests: Silence some intentional errors

Also fix printing messages in ClassDB test.

aaronfranke committed:
Check if the axis is zero / vectors are colinear in Vector3 slerp

YuriSizov committed:
Add text overrun behavior for the theme type selector

akien-mga committed:
Force disable S3TC support on Android/iOS since we don't handle it

Fixes #63909 for now.
This could be improved in the future if we want to properly support S3TC on mobile.

Faless committed:
[Scons] Implement module dependency sorting.

Modules can now call:

env.module_add_dependencies(name: str, deps: list, optional: bool)

To add required or optional dependencies during the "can_build" step.

Required dependencies will be checked and the module will be not be
enabled when they are missing, printing a warning to notify the user.

rburing committed:
Fix SoftDynamicBody3D normals

Store normal vector in A2B10G10R10 format.

akien-mga committed:
Merge pull request #63918 from akien-mga/no-s3tc-for-mobile

akien-mga committed:
Merge pull request #63915 from YuriSizov/editor-theme-type-is-wideboi

asmaloney committed:
[make_rst] Remove (fake) return types on annotations

Based on discussion in #63561, Yuri mentioned that the return types are not necessary here, so remove the fake "void" return types on annotations.

bruvzg committed:
Fix overrun flag check and HarfBuzz safe to break flag check.

akien-mga committed:
Merge pull request #63852 from fire-forge/default-colors

akien-mga committed:
Merge pull request #61856 from iwek7/label_overflowing_text

YeldhamDev committed:
Fix some bugs with `Window`s

CristianoSimao committed:
doc: Clarify `MultiMesh.set_instance_color` re: white albedo color

If the user does not set the albedo color to pure white, then the
multiplication with a value different from 1 will produce modulated colors.

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

YuriSizov committed:
Round icon position and size in buttons to make them look sharper

akien-mga committed:
Merge pull request #63912 from aaronfranke/vec3-slerp-colinear

akien-mga committed:
Merge pull request #63924 from bruvzg/ts_reg_fixes

akien-mga committed:
Merge pull request #63922 from rburing/fix_softbody_normals

akien-mga committed:
Merge pull request #59844 from Calinou/rename-shader-param-methods

smix8 committed:
Add NavigationMesh baking for HeightMapShape

Adds HeightMapShape for StaticColliders and GridMap to the NavigationMeshGenerator for baking NavigationMesh.

akien-mga committed:
Sync controller mappings DB with SDL2 community repo

Synced with gabomdq/SDL_GameControllerDB@4896d2de6bac388b5f8f8a42d0c39c5892bd5847

akien-mga committed:
certs: Sync with Mozilla bundle as of Jul 19, 2022

https://github.com/bagder/ca-bundle/commit/7f33e7eb8472dbcf31fdcf50cd216c89a282825d

akien-mga committed:
i18n: Sync editor translations with Weblate

(cherry picked from commit 4e64c1f1b6c842cbf343f4f1f176f6fa0671bbc7)

akien-mga committed:
i18n: Sync classref translations with Weblate

(cherry picked from commit b671ea88fb71c93213a08adcb568935cca0dae45)

akien-mga committed:
Merge pull request #63930 from akien-mga/sdl-gamecontrollerdb

akien-mga committed:
Merge pull request #63931 from akien-mga/certs-update

certs: Sync with Mozilla bundle as of Jul 19, 2022

Calinou committed:
Improve `range_lerp()` and related methods documentation

akien-mga committed:
Merge pull request #63916 from akien-mga/tests-silence-some-intentional-errors

Tests: Silence some intentional errors

Faless committed:
[Net] IP uses print_verbose when getaddrinfo fails.

Avoid spamming errors when network is disconnected.
Returned address will be invalid, so it can be checked by the user via
`ret.is_valid_ip_address`.

reduz committed:
Add a Framebuffer cache

Adds a FramebufferCache singletion that operates the same way as UniformSetCache.

Allows creating framebuffers on the fly (and keep them cached if re-requested) such as:

```C++
RID fb = FramebufferCache::get_singleton()->get_cache(texture1,texture2);
```

akien-mga committed:
Merge pull request #55452 from Calinou/editor-theme-tweak-popup-spacing

akien-mga committed:
Merge pull request #50297 from Calinou/distance-fade-use-interleaved-gradient-noise

bruvzg committed:
Implement `screen_is_kept_on` for macOS.

akien-mga committed:
Merge pull request #63941 from Calinou/doc-range-lerp

Improve `range_lerp()` and related methods documentation

Faless committed:
Merge pull request #63944 from Faless/net/4.x_ip_no_err

[Net] IP uses print_verbose when getaddrinfo fails.

akien-mga committed:
Merge pull request #63923 from asmaloney/doc-annotation-return-type

hakro committed:
Remove rect_ prefix from control properties when keyframing

clayjohn committed:
Merge pull request #63951 from reduz/framebuffer-cache

Add a Framebuffer cache

bruvzg committed:
[Text Server] Prevent composite glyphs which incorporate kashida from being used for justification. Update TextServer tests to clean up on fail.

YuriSizov committed:
Rebuild the trees in the EditorFeatureProfile dialog when the editor theme changes

akien-mga committed:
Merge pull request #63957 from YuriSizov/buttons-in-focus

Round icon position and size in buttons to make them look sharper

qarmin committed:
Check also GLES3 in CI

TomBedfordCodes committed:
updated the String class with info about the 'allow_empty' optional argument

akien-mga committed:
Merge pull request #63963 from YuriSizov/editor-feature-profiles-like-it-dark-and-damp

Rebuild the trees in the `EditorFeatureProfile` dialog when the editor theme changes

akien-mga committed:
Merge pull request #63953 from bruvzg/mac_keep_on_read

Implement `screen_is_kept_on` for macOS.

akien-mga committed:
Merge pull request #63882 from RedMser/keep-screen-on-windows

Implement `keep_screen_on` for Windows

aaronfranke committed:
Make "Godot source files" comment consistent in modules

akien-mga committed:
Merge pull request #63962 from bruvzg/fix_elong_liga

[Text Server] Prevent composite glyphs which incorporate kashida from being used for justification.

akien-mga committed:
Merge pull request #63958 from hakro/fix-control-animation-keys

Remove rect_ prefix from control properties when keyframing

akien-mga committed:
Merge pull request #63887 from qarmin/more_renames

Various converter enhancements and bugfixes

akien-mga committed:
Merge pull request #63871 from bruvzg/ts_build_options

Add Text Server related options to the build profiles editor.

akien-mga committed:
Merge pull request #63859 from Rindbee/fix-hue-slider-incorrectly-rotated

Fix Hue slider incorrectly rotated 90 degrees in ColorPicker

akien-mga committed:
Merge pull request #63776 from fire-forge/shapecast2d

Add ShapeCast2D editor handle and improve debug drawing

akien-mga committed:
Merge pull request #63762 from bruvzg/ft_warn

[TextServer] Add a build warning when building with external FreeType without Brotli support.

akien-mga committed:
Merge pull request #63704 from fire-forge/no-code

Fix EditorPropertyMultilineText expression mode

akien-mga committed:
Merge pull request #62787 from RandomShaper/vk_object_name

Set default resource names under pure debug in Vulkan RD

akien-mga committed:
Merge pull request #63698 from PrecisionRender/plane-operator-asterisk

Add operator `*` to `Plane`

akien-mga committed:
Merge pull request #63700 from YeldhamDev/no_methods_to_filter_pal

Hide method filter bar when selecting a doc in the Script editor

lawrnce committed:
Remove unnecessary string replacement

lawrnce committed:
Merge branch 'master' into bugfix-ios-export

YuriSizov committed:
Add checks for empty/unnamed arguments to make_rst.py

Faless committed:
[Websocket] Remove dummy JS server implemenation.

It does nothing but printing errors (only websocket client is available
in browsers), so we might as well return null when you instantiate it.

raulsntos committed:
Add conversion for `sync` and `slave` GDScript keywords

raulsntos committed:
Add conversion for the RPC attributes in C#

akien-mga committed:
Merge pull request #62847 from RandomShaper/upgrade_spirv_reflect

spirv-reflect: Update to latest commit

akien-mga committed:
Merge pull request #63870 from Calinou/doc-editorsettings

akien-mga committed:
Merge pull request #63977 from TomBCodes/master

akien-mga committed:
Merge pull request #63976 from aaronfranke/godots-own-comment-lines

akien-mga committed:
Merge pull request #63712 from object71/fix-export-issues

akien-mga committed:
Merge pull request #63985 from raulsntos/convert-rpc-attribute-csharp

RPC improvements to the conversion tool

YuriSizov committed:
Add support for [param foo] syntax in valid documentation contexts

YuriSizov committed:
Rename the argument tag to param in XML documentation

akien-mga committed:
Merge pull request #63992 from qarmin/check_also_GLES3

Test also GLES3 in CI

akien-mga committed:
Merge pull request #60843 from hoontee/CSG3

Add `TorusMesh`

RedMser committed:
Only define `keep_screen_on` project setting once

akien-mga committed:
Merge pull request #60924 from KoBeWi/derp

akien-mga committed:
Merge pull request #63361 from KoBeWi/floorf_lol

akien-mga committed:
Merge pull request #60567 from Geometror/make-unique-recursive

akien-mga committed:
Merge pull request #60299 from Calinou/editor-vram-compress-pc-improve-warning

kleonc committed:
`FileSystemDock` Fix crash when renaming a file in split mode

timothyqiu committed:
Use %s for bool value in vformat

Calinou committed:
Fix broken rST reference in HTTPClient documentation

References in reStructuredText cannot be located within bold/italic
text. This caused the online class reference to have the `:ref:`
text written as-is.

DarkKilauea committed:
Fix NavigationRegion3D gizmo's odd visual behavior

akien-mga committed:
Merge pull request #64031 from Calinou/doc-httprequest-no-bold-ref

Fix broken rST reference in HTTPClient documentation

akien-mga committed:
Merge pull request #64025 from kleonc/filesystem-dock-rename-crash-fix

`FileSystemDock` Fix crash when renaming a file in split mode

akien-mga committed:
Merge pull request #64028 from timothyqiu/input-event-bool

Use `%s` for bool value in vformat

akien-mga committed:
Merge pull request #61893 from Calinou/light3d-add-shadow-opacity

Add a shadow opacity property to Light3D

Geometror committed:
Vector4/Vector4i: Add missing methods, tests and fix change of sign operator

touilleMan committed:
Expose ObjectID among the native structure in GDExtension

AntonioDell committed:
fix(gdscript): Infer type from preload const

When resolving the type of the attribute from the variant, the result_type.kind was overritten for no reason.
It is assumed that this only needs to be done, if the variant value is not valid to have any kind here.

Solves #63715

Faless committed:
[MP] Cleanup and fixes for replication plugin.

- Remove dead code.
- Fix "Add from path" adding the wrong string when targeting root node.

Faless committed:
Merge pull request #64047 from Faless/ws/4.x_no_emws

[Websocket] Remove dummy JS server implemenation.

pulawskig committed:
Fix axis snapping of CollisionPolygon2d's newly created vertex

akien-mga committed:
Merge pull request #58494 from rcorre/recent-projects

akien-mga committed:
Merge pull request #62089 from Step-an/add_drag_and_drop_to_EditorPropertyPath

akien-mga committed:
Merge pull request #64049 from Faless/mp/4.x_repl_editor_cleanup

akien-mga committed:
Merge pull request #64027 from Geometror/add-vector4-tests

akien-mga committed:
Merge pull request #63919 from Faless/scons/4.x_easy_deps

akien-mga committed:
Merge pull request #64045 from touilleMan/gdextension-ObjectID-in-native_structures

Geometror committed:
Allow vec2 and vec4 for reflect and refract

akien-mga committed:
Merge pull request #62454 from Geometror/reflect-refract-vec2-vec4

Allow for vec2/vec4 to be used in reflect and refract

raulsntos committed:
Various fixes to C# documentation

DarkKilauea committed:
Fix typo when setting NavigationRegion travel_cost

Calinou committed:
Remove FOV adjustment with Alt + mouse wheel in the 3D editor

This shortcut got in the way when using the Maya navigation scheme,
and also when using the slow freelook modifier (regardless of
navigation scheme).

akien-mga committed:
Merge pull request #64075 from raulsntos/csharp-doc-fixes

Various fixes to C# documentation

Nosliwnayr committed:
Add area monitor callback error checking

raulsntos committed:
Avoid paths with invalid characters in `IsRider`

akien-mga committed:
Merge pull request #64068 from DarkKilauea/fix-region-travel-cost

akien-mga committed:
Merge pull request #64046 from AntonioDell/bugfix/63715-infer-preloaded-const-types

akien-mga committed:
Merge pull request #64085 from raulsntos/fix-rider-check

akien-mga committed:
Merge pull request #64050 from pulawskig/master

Commit:
capitalize shader uniform groups

akien-mga committed:
Merge pull request #63926 from CristianoSimao/patch-1

akien-mga committed:
Merge pull request #62571 from Calinou/gpuparticlescollisionsdf3d-add-bake-mask

akien-mga committed:
Merge pull request #62547 from clayjohn/ref_vec_pbr

akien-mga committed:
Merge pull request #64093 from nathanfranke/shader-capitalize

akien-mga committed:
Merge pull request #63980 from lawrnce/bugfix-ios-export

akien-mga committed:
Merge pull request #63925 from YeldhamDev/window_fixes

akien-mga committed:
Merge pull request #64070 from Calinou/3d-editor-remove-alt-fov-adjustment

A-Lamia committed:
fix: modifies workspace->root_uri so that p_uri is symmetrical to other operating systems. #63388

akien-mga committed:
Merge pull request #64014 from RedMser/keep-screen-on-singledef

akien-mga committed:
Merge pull request #62868 from fire-forge/array-resize-size

Use a SpinBox in the inspector array resize dialog

akien-mga committed:
Merge pull request #62861 from samdze/image-size-vector2i

Make Image.get_size() return a Vector2i instead of a Vector2

YuriSizov committed:
Add tests for empty/unnamed arguments to ClassDB, Variant, GDScript

and3rson committed:
Add conversions for tweens, fix move_and_slide/start/await conversions

akien-mga committed:
Merge pull request #63173 from fire-forge/layers-grid

akien-mga committed:
Merge pull request #64078 from Nosliwnayr/master

akien-mga committed:
Merge pull request #63932 from smix8/navigation_heightmapshape_baking_4.x

akien-mga committed:
Merge pull request #62885 from madmiraal/fix-59931

akien-mga committed:
Merge pull request #63518 from timothyqiu/window-toggle

akien-mga committed:
Merge pull request #63253 from V-Sekai/disable_sprite_conversion_on_foreign_nodes

bruvzg committed:
Fix macOS and iOS defines in the rendering code.

antonWetzel committed:
vector4 distance_squared_to and update csharp

akien-mga committed:
Merge pull request #64104 from and3rson/convert3to4-fixes-2

akien-mga committed:
Merge pull request #64004 from YuriSizov/doctool-fail-on-unnamed-args

Add checks and tests for empty/unnamed arguments

akien-mga committed:
Merge pull request #64076 from raulsntos/doc-fixes

Add `@GDScript.type_exists` documentation, Fix typo in weakref documentation

akien-mga committed:
Merge pull request #63632 from dsnopek/fix-locale-remap-with-binary-resources-4.x

[4.x] Fix locale resource remapping with binary conversion on export

akien-mga committed:
Merge pull request #64106 from bruvzg/fix_os_names

akien-mga committed:
Merge pull request #63258 from Calinou/gdextension-print-expected-config-os-arch

akien-mga committed:
Merge pull request #64034 from DarkKilauea/fix-nav-region-gizmo

akien-mga committed:
Merge pull request #64008 from YuriSizov/doctool-add-param-reference-syntax

akien-mga committed:
Merge pull request #64099 from A-Lamia/lsp_uri_path_fixes

Calinou committed:
Improve InstancePlaceholder documentation

akien-mga committed:
Merge pull request #64127 from Calinou/doc-instanceplaceholder

asmaloney committed:
[doc] Use "param" instead of "code"  to refer to parameters

Co-authored-by: Yuri Sizov <11782833+YuriSizov@users.noreply.github.com>

PrinceDeveloperOf committed:
Added to Sprite3D Documentation

BastiaanOlij committed:
For dev builds, keep track of resource names in the Vulkan driver

BastiaanOlij committed:
Fix resource leaks in VoxelGI

akien-mga committed:
Merge pull request #64139 from BastiaanOlij/fix_gi_resource_leaks

Fix resource leaks in VoxelGI

akien-mga committed:
Merge pull request #64135 from PrinceDeveloperOf/master

Added to Sprite3D Documentation

smix8 committed:
Fix Editor Navigation debug edge connection visuals

Fixes missing Navigation debug edge connection visuals in Editor due to disabled NavigationServer.

bruvzg committed:
Fix memory leak when accessing/listing system fonts.

smix8 committed:
Add NavigationServer2D debug functions

Adds debug functions to NavigationServer2D to mirror NavigationServer3D functions for 2D users.

Calinou committed:
Tweak Label3D defaults for better readability

- Increase font size and decrease pixel size.
  - The font is rendered at the same physical size, but is more detailed,
    which is visible when the camera is up close.
- Add an outline to improve readability on mixed-color backgrounds.
  - The outline is fairly thick to ensure it doesn't get too grainy
    at a distance (without requiring MSDF or mipmaps on the default
    project font).

hakro committed:
Prevent AnimationPlayer from being added on GLTF import if the option is unchecked. Fixes #63954

akien-mga committed:
Merge pull request #64137 from BastiaanOlij/dev_track_resource_names

RandomShaper committed:
Apply correct formatting to comments in the Vulkan driver

akien-mga committed:
Merge pull request #64134 from asmaloney/doc-params

[doc] Use "param" instead of "code"  to refer to parameters

akien-mga committed:
Merge pull request #64153 from smix8/navigation_add_2dserver_debug_functions_4.x

akien-mga committed:
Merge pull request #64146 from smix8/fix_editor_disabled_navigationserver_4.x

akien-mga committed:
Merge pull request #64151 from bruvzg/fc_leak_fix

YeldhamDev committed:
Fix error when switching to another `GridMap` with an item with higher index selected

akien-mga committed:
Merge pull request #64108 from antonWetzel/csharp-update

danboo committed:
Add note about using AudioListener2D

There was no mention of the effect of AudioListener2D in this documentation, making it unclear if there was a relationship. The new text is copied and modified from the AudioStreamPlayer3D documentation.

Use 'an' as article before 'AudioListener3D'

TokageItLab committed:
add get_depth() to KinematicCollision3D as same of 2D

akien-mga committed:
Merge pull request #62722 from aaronfranke/there-can-be-only-one

Remove duplicate data structure for editor export features

Chaosus committed:
Prevent global functions from overriding completion of subscript

timothyqiu committed:
Expose clear method for packed arrays

akien-mga committed:
Merge pull request #64178 from RandomShaper/fix_vk_comments

akien-mga committed:
Merge pull request #64195 from TokageItLab/add-get-depth-3d

akien-mga committed:
Merge pull request #64160 from Calinou/tweak-label3d-defaults

akien-mga committed:
Merge pull request #64188 from YeldhamDev/gridmap_item_fix

akien-mga committed:
Merge pull request #64206 from Chaosus/gds_fix

akien-mga committed:
Merge pull request #64192 from danboo/danboo-mention-audiolistener2d

akien-mga committed:
Merge pull request #64208 from timothyqiu/packed-clear

akien-mga committed:
Merge pull request #62925 from aaronfranke/pck-export

akien-mga committed:
Merge pull request #49829 from naithar/feature/ios-plugins-swift-4.0

akien-mga committed:
Merge pull request #64169 from hakro/fix-gltf-animations-always-imported

akien-mga committed:
Merge pull request #63245 from V-Sekai/animation_editor_read_only

nonunknown777 committed:
implement Array.pick_random()

mhilbrunner commented:
Seems this includes a lot of other commits :) Needs a proper rebase.

akien-mga commented:
@nonunknown has been struggling with rebasing on this one. Maybe a contributor can help salvage this?
